include/rapidcheck/state/gen/ParallelCommands.hpp

```diff
@@ -0,0 +1,341 @@
+#pragma once
+
+#include "rapidcheck/state/gen/Commands.h"
+
+namespace rc {
+namespace state {
+namespace gen {
+
+namespace detail {
+template <typename Cmd, typename GenFunc>
+class ParallelCommandsGen {
+public:
+  using CmdSP = std::shared_ptr<const Cmd>;
+  using Model = typename Cmd::Model;
+  using Sut = typename Cmd::Sut;
+
+  template <typename ModelArg, typename GenFuncArg>
+  ParallelCommandsGen(ModelArg &&initialState, GenFuncArg &&genFunc)
+      : m_initialState(std::forward<ModelArg>(initialState))
+      , m_genFunc(std::forward<GenFuncArg>(genFunc)) {}
+
+  Shrinkable<ParallelCommands<Cmd>> operator()(const Random &random,
+                                               int size) const {
+    return generateCommands(random, size);
+  }
+
+private:
+  struct CommandEntry {
+    CommandEntry(Random &&aRandom, Shrinkable<CmdSP> &&aShrinkable)
+        : random(std::move(aRandom))
+        , shrinkable(std::move(aShrinkable)) {}
+
```

Comment:
Maybe we really do should keep track of the state in between commands if we also want to pass that as a hint even for parallel commands? The question is which state. In the prefix, it is simple but what about the parallel? Should you always pass the state as it looks like after the prefix? Or treat the branches as independent and pass the state as it looks like after only the commands in that particular parallel branch have been executed?